core/src/main/java/org/sonarsource/sonarlint/core/SonarLintClientImpl.java

```diff
@@ -94,16 +96,35 @@ private void checkStarted() {
 
   private static RuntimeException handleException(RuntimeException t) {
     if (LOG.isDebugEnabled()) {
-      return t;
+      // In DEBUG mode always return full exception
+      return convertToSonarLintException(t);
     }
 
     for (Throwable y : Throwables.getCausalChain(t)) {
       if (y instanceof MessageException) {
-        return (MessageException) y;
+        return convertToSonarLintException(y);
       }
     }
 
-    return t;
+    return convertToSonarLintException(t);
+  }
+
+  private static RuntimeException convertToSonarLintException(@Nullable Throwable t) {
+    if (t == null) {
+      return null;
+    }
+    Throwable cause = convertToSonarLintException(t.getCause());
+    if (t instanceof RuntimeException && isPartOfJDK(t)) {
```

Comment:
You're right, this is buggy. And since I don't think I can change the cause of a RuntimeException to replace it with a SonarLintException in case it is not part of the JDK, I will simply convert all the stack to SonarLintException instances.